compiler/rustc_data_structures/src/binary_search_util/mod.rs

```diff
@@ -14,31 +14,18 @@ where
     let start = data.partition_point(|x| key_fn(x) < *key);
     // At this point `start` either points at the first entry with equal or
     // greater key or is equal to `size` in case all elements have smaller keys
+    // Invariant: start == size || key_fn(&data[start]) >= *key
```

Comment:
I'm not sure what the `// Invariant: ` comments add. The plain english comments are useful though.